### DISCUSSION TITLE
Notifications configuration help

### DISCUSSION THREAD

**Discussion by eliduke** [juftin/camply#32] [Q&A] [answered]
I am currently setting up a simple EC2 instance to run camply continuously and get notifications when things open up.

I'm looking at the example notification configuration file [here](https://github.com/juftin/camply/blob/main/docs/examples/example.camply):

```
# ANY VARIABLES THAT ARE SET TO "" ARE REQUIRED, VARIABLES THAT HAVE A
# VALUE PROVIDED ARE THE DEFAULT AND CAN BE CHANGED OR REMOVED ENTIRELY

# REQUIRED TO SEND PUSHOVER NOTIFICATIONS
PUSHOVER_PUSH_TOKEN=""
PUSHOVER_PUSH_USER=""

# REQUIRED TO SEND PUSHBULLET NOTIFICATIONS
PUSHBULLET_API_TOKEN=""

# REQUIRED TO SEND EMAIL NOTIFICATIONS
EMAIL_TO_ADDRESS=""
EMAIL_USERNAME=""
EMAIL_PASSWORD=""
EMAIL_SMTP_SERVER="smtp.gmail.com"
EMAIL_SMTP_PORT=465
EMAIL_FROM_ADDRESS="camply@juftin.com"
EMAIL_SUBJECT_LINE="Camply Notification"
```

And I'm a little bit confused on a couple of these options and what the expected value is.

**PUSHOVER**

To start with, I signed up for Pushover, but I'm not sure exactly what the two different keys are. When I signed up for pushover, it immediately showed me a random string that it referred to as the "user key", but I can't seem to find a "token" anywhere in there. So, am I missing something? Any direction there would be helpful.

**PUSHBULLET**

I signed up for pushbullet, but then it says that the iPhone app is not currently available in my region. Maybe it's android only now? Anyone else seeing something different?

**EMAIL**

In the end, I most-likely will just use email, so I tried configuring that, but I'm getting this error when I run camply with continuous notifications set to email:

```
raise SMTPAuthenticationError(code, resp)
smtplib.SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials h9sm13883702qkj.66 - gsmtp')
```

And then I got this email:

<img width="530" alt="Screen Shot 2021-06-30 at 4 04 44 PM" src="https://user-images.githubusercontent.com/1372520/124043316-a11dd580-d9bf-11eb-9cdb-4ac02365db02.png">

Some of the documentation I'm seeing is talking about "app passwords", but it also is sorta NOT RECOMMENDING them.

Anyone have any input on this?

Here is the full stack trace on the error from above:

```
Traceback (most recent call last):
  File "/home/ec2-user/.local/bin/camply", line 8, in <module>
    sys.exit(main())
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/utils/camply_cli.py", line 39, in main
    camply_cli.run_cli()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/utils/camply_cli.py", line 418, in run_cli
    self.execute_cli_arguments()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/utils/camply_cli.py", line 338, in execute_cli_arguments
    self.execute_campsites()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/utils/camply_cli.py", line 405, in execute_campsites
    camping_finder.get_matching_campsites(**search_kwargs)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/search/base_search.py", line 351, in get_matching_campsites
    search_forever=search_forever)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/search/base_search.py", line 303, in _search_campsites_continuous
    continuous_search_attempts=continuous_search_attempts)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/search/base_search.py", line 229, in _continuous_search_retry
    SilentNotifications)()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/camply/notifications/email_notifications.py", line 60, in __init__
    password=EmailNotifications._email_password)
  File "/usr/lib64/python3.7/smtplib.py", line 730, in login
    raise last_exception
  File "/usr/lib64/python3.7/smtplib.py", line 721, in login
    initial_response_ok=initial_response_ok)
  File "/usr/lib64/python3.7/smtplib.py", line 642, in auth
    raise SMTPAuthenticationError(code, resp)
smtplib.SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials g6sm5927400qtk.44 - gsmtp')
```

**Comment by eliduke** [comment]
I ended up giving up on gmail and configured my fastmail account, which seems to be working so far. I've got camply running continuously in EC2 (with tmux), and I'll check back here there to see how it holds up!

**Comment by juftin** [comment, marked as answer]
Thanks so much for the detailed questions @eliduke, that really makes this useful for other users too.

## Pushover:
You will need to include your Pushover API token and your Pushover User token. As far as I know, only the user token is accessible via the app whereas the API token needs to be linked to a "pushover application" from Pushover's website here: https://pushover.net/apps/build

I should probably just create and host an official Camply App on Pushover and embed it into the application. That way you will only need a user token. It's not free once you hit a certain volume, but I've had a couple of kind donors already and nowhere to spend the money so maybe that's a good idea. 

## Pushbullet:
~I'm actually not sure what's happening there, I was sure Pushbullet supports iPhone and Android~ Pushbullet supports Android, Windows, Mac, and Web Browser - sorry about that 

## Email

I think you have may have MFA turned on for your Google Account - so username/password auth isn't working. To get around this you need to create an [App Password](https://support.google.com/accounts/answer/185833?hl=en) with Google. App passwords basically work everywhere - so keep it safe.


Good luck, I hope you get your site!

**Reply by dallasg89** [reply]
I can't for the life of me get the notifications running. Everything else has gone smooth, can search for available sites and get results. But when I run it with notifications nothing ever goes through. Have set up and configured pushover, pushbullet and my email and none work. Done the camply configure to setup all of the above and checked the file and everything looks good.

I can run it with a known bad password on the email and it doesn't even return an error like the above person. I've got to be missing something but I just can't figure it out. 

Love the idea of this and would love to use it for my local campground which is always booked on weekends. 

Any idea what I'm doing wrong. Or any steps you can point me towards figuring it out. Brand new to python by the way, so that's probably the root of the cause.

**Reply by juftin** [reply]
@dallasg89 can you share the error that you're getting when trying to send notifications? The below command is a sample you can use (just a random campground that will send a notification on the first try)
```
camply campsites \
    --campground 232280 \
    --start-date 2021-10-01 \
    --end-date 2021-10-02 \
    --continuous \
    --notify-first-try \
    --notifications email
```

**Comment by dallasg89** [comment]
I don't actually get any error, I just never receive a notification. I also
don't get any error message even when I put in a bad email.

**Reply by juftin** [reply]
hmmm... Even more mysterious... can you share the command you're running too? what happens when you run the following command:
```
camply campsites \
    --campground 232280 \
    --start-date 2021-10-01 \
    --end-date 2021-10-02 \
    --continuous \
    --notify-first-try \
    --notifications email
 ```
 
 Please include all output from camply (feel free to anonymize your email). Here's an example from mine:
 ```
 ❯ camply campsites \
    --campground 232280 \
    --start-date 2021-10-01 \
    --end-date 2021-10-02 \
    --continuous \
    --notify-first-try \
    --notifications email
2021-08-11 18:42:32,028 [  CAMPLY]: camply, the campsite finder ⛺️
2021-08-11 18:42:32,035 [    INFO]: 1 booking nights selected for search, ranging from 2021-10-01 to 2021-10-01
2021-08-11 18:42:33,771 [    INFO]: 1 Matching Campgrounds Found
2021-08-11 18:42:33,771 [    INFO]: ⛰  Arapaho & Roosevelt National Forests Pawnee NG, CO (#1051) - 🏕  Kelly Dahl (#232280)
2021-08-11 18:42:39,004 [    INFO]: Searching for campsites every 10 minutes. Notifications active via <EmailNotifications>
2021-08-11 18:42:39,004 [    INFO]: Searching across 1 campgrounds
2021-08-11 18:42:39,004 [    INFO]: Searching Kelly Dahl, Arapaho & Roosevelt National Forests Pawnee NG, CO (232280) for availability: October, 2021
2021-08-11 18:42:39,189 [    INFO]: 	⛺️	200 total sites found in month of October
2021-08-11 18:42:39,366 [    INFO]: ⛺️ ⛺️ ⛺️ ⛺️ 8 Reservable Campsites Matching Search Preferences
2021-08-11 18:42:39,370 [    INFO]: 📅 Fri, October 01 🏕  8 sites
2021-08-11 18:42:39,371 [    INFO]: 	⛰️  Arapaho & Roosevelt National Forests Pawnee NG, CO  🏕  Kelly Dahl: ⛺ 8 sites
2021-08-11 18:42:39,372 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54001 (1 night)
2021-08-11 18:42:39,372 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54089 (1 night)
2021-08-11 18:42:39,372 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54201 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54262 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54439 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54458 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54900 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 		🔗 https://www.recreation.gov/camping/campsites/54913 (1 night)
2021-08-11 18:42:39,373 [    INFO]: 8 New Campsites Found.
2021-08-11 18:42:44,585 [    INFO]: Sending Email to juftin@juftin.com: Camply Notification
2021-08-11 18:42:45,418 [    INFO]: Email sent successfully
2021-08-11 18:42:45,445 [  CAMPLY]: Exiting camply 👋
```

**Comment by dallasg89** [comment]


**Reply by juftin** [reply]
Ahh @dallasg89 , there is a solution! Notifications only work when the `--continuous` flag is is set. The [Arguments Section](https://github.com/juftin/camply/blob/main/README.md#arguments) of the documentation has some more detail about it. 

camply (and most other command line interfaces) also has a `--help` parameter that will give you some pointers on the parameters and usage too. So for example, `camply --help` will give you info on the main camply utility while `camply campsites --help` will give you even more specific info on the campsite searching functionality. Also make sure to check on the `--notify-first-try` option as well as well as the `--search-forever` flags as they are very helpful for notifications too:
```
❯ camply campsites --help
2021-08-12 10:58:12,934 [  CAMPLY]: camply, the campsite finder ⛺️
usage: camply campsites [-h] [--rec-area RECREATION_AREA_ID] [--campground CAMPGROUND_ID] [--start-date START_DATE] [--end-date END_DATE] [--weekends] [--nights NIGHTS] [--provider PROVIDER] [--continuous]
                        [--polling-interval POLLING_INTERVAL] [--notifications NOTIFICATIONS] [--notify-first-try] [--search-forever] [--yml-config YML_CONFIG]

Search for a campsite within camply. Campsites are returned based on the search criteria provided. Campsites contain properties like booking date, site type (tent, RV, cabin, etc), capacity, price, and a
link to make the booking. Required parameters include `--start-date`, `--end-date`, `--rec-area` / `--campground`. Constant searching functionality can be enabled with `--continuous` and notifications can
be enabled using `--notifications`.

optional arguments:
  -h, --help            show this help message and exit
  --rec-area RECREATION_AREA_ID
                        Add Recreation Areas (comprised of campgrounds) by ID.
  --campground CAMPGROUND_ID
                        Add individual Campgrounds by ID.
  --start-date START_DATE
                        (YYYY-MM-DD) Start of Search window. You will be arriving this day.
  --end-date END_DATE   (YYYY-MM-DD) End of Search window. You will be checking out this day.
  --weekends            Only search for weekend bookings (Fri/Sat nights).
  --nights NIGHTS       Search for campsite stays with consecutive nights. Defaults to 1 which returns all campsites found.
  --provider PROVIDER   Camping Search Provider. Options available are 'Yellowstone' and 'RecreationDotGov'. Defaults to 'RecreationDotGov', not case-sensitive.
  --continuous          Continuously check for a campsite to become available, and quit once at least one campsite is found.
  --polling-interval POLLING_INTERVAL
                        If --continuous is activated, how often to wait in between checks (in minutes). Defaults to 10, cannot be less than 5.
  --notifications NOTIFICATIONS
                        If --continuous is activated, types of notifications to receive. Options available are 'email', 'pushover', 'pushbullet', or 'silent'. Defaults to 'silent' - which just logs messages
                        to console.
  --notify-first-try    If --continuous is activated, whether to send all non-silent notifications if more than 5 matching campsites are found on the first try. Defaults to false which only sends the first
                        5.
  --search-forever      If --continuous is activated, this method continues to search after the first availability has been found. The one caveat is that it will never notify about the same identical
                        campsite for the same booking date.
  --yml-config YML_CONFIG
                        Rather than provide arguments to the command line utility, instead pass a file path to a YAML configuration file. See the documentation for more information on how to structure your
                        configuration file.
2021-08-12 10:58:12,940 [  CAMPLY]: Exiting camply 👋
```

Let me know how it goes, hopefully this should get you running. Good luck on your search!

**Reply by dallasg89** [reply]
That seems like it is it. I don't know how I missed the "if --continuous is activate..." when reading through it. 

Works beautifully now, thank!

Just have one last question, and I don't know if this is even possible. Is there some way I can modify the links scraped from recreation.gov to open inside the recreation.gov app? On android by the way.

**Reply by dsonniesideup** [reply]
Glad I saw this comment cause I just spend the last 1.5 hrs trying to figure out why notification wasn't working. Missed the --continuous

**Comment by javagenki** [comment]
I've managed to get pushover and email working but separately. When I try to combine them in a yaml file, neither of them work and no error. I've tried the following formats:

notifications:    pushover, email
notifications:    (pushover, email)

**Reply by juftin** [reply]
Hey @javagenki, `camply` isn't currently configured to notify different providers on the same search - it's only limited to one. That is an interesting idea though, I'll give it some thought about how I might implement that

**Reply by juftin** [reply]
It tries to match a string (`pushover, email` in this case) to one of the qualifying notifier classes: https://github.com/juftin/camply/blob/3542344723f03e15d7bb8aeed272dfe6357c0d38/camply/notifications/__init__.py#L14-L19

If no match is found it defaults to the `SilentNotifications` class which just logs to console - for now you will need to choose one or the other

🤔 Here is where the code I'd need to adapt to accept a list: https://github.com/juftin/camply/blob/3542344723f03e15d7bb8aeed272dfe6357c0d38/camply/search/base_search.py#L200-L264

I could probably build out the functionality by next week

**Reply by javagenki** [reply]
Wow, thank you for the speedy reply and thank you so much for creating Camply! I'm sorry I didn't realize that it was limited to one. One should be enough. I think I wanted to be doubly sure that I didn't miss that notification. :)
